File: src/core/lexer.ts
/**
 * @fileoverview CodonCanvas genome lexer
 *
 * Tokenizes DNA/RNA-like triplet syntax into executable codon tokens.
 * Supports both DNA (T) and RNA (U) notation with automatic normalization.
 *
 * **Parsing Rules:**
 * - Valid bases: A, C, G, T, U (U is normalized to T)
 * - Whitespace is ignored (used for formatting)
 * - Comments start with `;` and extend to end of line
 * - Source length must be divisible by 3 (complete codons only)
 *
 * @module core/lexer
 */

import {
  BASES,
  type Codon,
  type CodonToken,
  type GenomeMetadata,
  type ParseError,
  type ValueMode,
} from "@/types";

/** Valid base letters for lexer validation, derived from BASES */
const VALID_BASES = new Set<string>(Object.keys(BASES));

/**
 * Strip comment from a line (everything after `;`).
 * @internal
 */
function getCodeContent(line: string): string {
  const commentIdx = line.indexOf(";");
  return commentIdx >= 0 ? line.slice(0, commentIdx) : line;
}

/**
 * Result of parsing a genome source, including tokens and metadata.
 */
export interface ParseResult {
  /** Parsed codon tokens */
  tokens: CodonToken[];
  /** Extracted genome metadata from directives */
  metadata: GenomeMetadata;
}

/**
 * Default genome metadata when no directives are specified.
 *
 * Default is "centered" mode for bidirectional positioning.
 * Use `; @mode: forward-only` for legacy compatibility with older genomes.
 */
export const DEFAULT_GENOME_METADATA: GenomeMetadata = {
  mode: "centered",
};

/**
 * Maps user-friendly mode names (used in genome files) to internal ValueMode.
 *
 * User-facing names in genome files:
 * - `bidirectional` → "centered" (can move in any direction)
 * - `forward-only` → "forward" (legacy, positive movement only)
 */
const MODE_ALIASES: Record<string, ValueMode> = {
  // User-friendly names
  bidirectional: "centered",
  "forward-only": "forward",
  // Also accept internal names for advanced users
  centered: "centered",
  forward: "forward",
};

/**
 * Lexer interface for CodonCanvas genome parsing.
 *
 * Responsible for tokenizing DNA/RNA triplets and validating genome structure.
 * Implementations should normalize RNA (U) to DNA (T) notation.
 */
export interface Lexer {
  /**
   * Tokenize source genome into codons.
   *
   * @param source - Raw genome string containing DNA/RNA bases (A/C/G/T/U)
   *   with optional whitespace and comments (`;` to end of line)
   * @returns Array of codon tokens with position and line information
   * @throws {Error} If invalid characters found (non-ACGTU after stripping whitespace)
   * @throws {Error} If source length not divisible by 3 (incomplete codon)
   */
  tokenize(source: string): CodonToken[];

  /**
   * Parse source genome into tokens and extract metadata directives.
   *
   * Directives are special comments that configure genome execution:
   * - `; @mode: signed` - Enable signed value interpretation
   * - `; @mode: unsigned` - Unsigned values (default)
   *
   * @param source - Raw genome string with optional directives
   * @returns Parsed tokens and extracted metadata
   * @throws {Error} If invalid characters or incomplete codons
   *
   * @example
   * ```typescript
   * const lexer = new CodonLexer();
   * const { tokens, metadata } = lexer.parse(`
   *   ; @mode: signed
   *   ATG GAA GGG ACA TAA
   * `);
   * // metadata.mode === "signed"
   * ```
   */
  parse(source: string): ParseResult;

  /**
   * Validate reading frame alignment.
   *
   * Checks for whitespace breaks within triplets that would disrupt the reading frame.
   * Does NOT throw; returns warnings for formatting issues.
   *
   * @param source - Raw genome string to validate
   * @returns Array of frame alignment warnings (never errors)
   */
  validateFrame(source: string): ParseError[];

  /**
   * Validate structural integrity of tokenized genome.
   *
   * Checks for:
   * - START codon (ATG) at beginning
   * - STOP codon (TAA/TAG/TGA) at end
   * - Unknown/unmapped codons
   * - Code after STOP (unreachable)
   *
   * @param tokens - Array of codon tokens to validate
   * @returns Array of structural errors and warnings
   */
  validateStructure(tokens: CodonToken[]): ParseError[];
}

/**
 * CodonCanvas lexer implementation.
 * Parses DNA/RNA-like triplet syntax into executable codon tokens.
 * Supports both DNA (T) and RNA (U) notation.
 *
 * @example
 * ```typescript
 * const lexer = new CodonLexer();
 * // DNA notation
 * const tokens1 = lexer.tokenize('ATG GAA CCC GGA TAA');
 * // RNA notation
 * const tokens2 = lexer.tokenize('AUG GAA CCC GGA UAA');
 * // Both produce identical results
 * const errors = lexer.validateStructure(tokens1);
 * ```
 */
export class CodonLexer implements Lexer {
  private readonly validBases = VALID_BASES;

  /**
   * Tokenize source genome into codons.
   *
   * Strips comments (`;` to EOL), removes whitespace, validates base characters,
   * and chunks into triplets. Tracks line numbers for error reporting.
   * Supports both DNA (T) and RNA (U) notation - U is normalized to T internally.
   *
   * @param source - Raw genome string (supports A/C/G/T/U, whitespace, `;` comments)
   * @returns Array of codon tokens with position metadata
   * @throws {Error} If invalid characters found (message includes line/column)
   * @throws {Error} If cleaned source length not divisible by 3
   *
   * @example
   * ```typescript
   * const lexer = new CodonLexer();
   * // DNA notation
   * const tokens1 = lexer.tokenize('ATG GAA CCC GGA TAA');
   * // RNA notation (U→T normalized)
   * const tokens2 = lexer.tokenize('AUG GAA CCC GGA UAA');
   * // Both produce identical results
   * ```
   */
  tokenize(source: string): CodonToken[] {
    // Track line numbers for error reporting
    const lines = source.split("\n");
    let cleanedSource = "";
    const positionMap: Array<{ line: number; column: number }> = [];

    // Strip comments and track positions
    for (let lineIdx = 0; lineIdx < lines.length; lineIdx++) {
      const line = lines[lineIdx];
      if (line === undefined) continue;
      this.processTokenizeLine(line, lineIdx, (char, pos) => {
        cleanedSource += char;
        positionMap.push(pos);
      });
    }

    // Check for non-triplet length
    if (cleanedSource.length % 3 !== 0) {
      throw new Error(
        `Source length ${cleanedSource.length} is not divisible by 3. Missing ${
          3 - (cleanedSource.length % 3)
        } bases for complete codon.`,
      );
    }

    // Chunk into triplets
    const tokens: CodonToken[] = [];
    for (let i = 0; i < cleanedSource.length; i += 3) {
      const codonText = cleanedSource.slice(i, i + 3);
      tokens.push({
        text: codonText as Codon,
        position: i,
        line: positionMap[i]?.line || 1,
      });
    }

    return tokens;
  }

  /**
   * Validate reading frame alignment.
   *
   * Detects whitespace breaks within triplets that would disrupt codon boundaries.
   * Useful for linting genome source code for readability vs correctness.
   *
   * @param source - Raw genome string to validate
   * @returns Array of frame alignment warnings
   *
   * @example
   * ```typescript
   * const lexer = new CodonLexer();
   * const errors = lexer.validateFrame('ATG GG A CCA'); // Mid-triplet break: "GG A"
   * // Returns: [{ message: "Mid-triplet break...", severity: 'warning', ... }]
   * ```
   */
  validateFrame(source: string): ParseError[] {
    const errors: ParseError[] = [];
    const lines = source.split("\n");

    for (let lineIdx = 0; lineIdx < lines.length; lineIdx++) {
      const line = lines[lineIdx];
      if (line === undefined) continue;
      this.validateLineFrame(line, lineIdx, errors);
    }

    return errors;
  }

  /**
   * Validate structural integrity of tokenized genome.
   *
   * Checks for:
   * - START codon (ATG) at beginning
   * - Proper STOP codon (TAA/TAG/TGA) placement
   * - Unknown/unmapped codons
   * - START after STOP warnings
   *
   * @param tokens - Array of codon tokens to validate
   * @returns Array of structural errors and warnings
   *
   * @example
   * ```typescript
   * const lexer = new CodonLexer();
   * const tokens = lexer.tokenize('TAA GGA CCA'); // Missing START, early STOP
   * const errors = lexer.validateStructure(tokens);
   * // Returns errors for missing START and STOP before execution
   * ```
   */
  validateStructure(tokens: CodonToken[]): ParseError[] {
    const errors: ParseError[] = [];

    if (tokens.length === 0) {
      return errors;
    }

    this.checkStartCodon(tokens, errors);
    this.checkStopCodons(tokens, errors);

    return errors;
  }

  private processTokenizeLine(
    line: string,
    lineIdx: number,
    onBase: (char: string, pos: { line: number; column: number }) => void,
  ) {
    const codeLine = getCodeContent(line);

    for (let charIdx = 0; charIdx < codeLine.length; charIdx++) {
      const char = codeLine[charIdx];
      if (char === undefined) continue;
      if (this.validBases.has(char)) {
        // Normalize U→T (RNA to DNA notation)
        const normalizedChar = char === "U" ? "T" : char;
        onBase(normalizedChar, { line: lineIdx + 1, column: charIdx });
      } else if (char.trim() !== "") {
        throw new Error(
          `Invalid character '${char}' at line ${
            lineIdx + 1
          }, column ${charIdx}`,
        );
      }
    }
  }

  /**
   * Parse source genome into tokens and extract metadata directives.
   *
   * @param source - Raw genome string with optional directives
   * @returns Parsed tokens and extracted metadata
   */
  parse(source: string): ParseResult {
    const tokens = this.tokenize(source);
    const metadata = this.extractDirectives(source);
    return { tokens, metadata };
  }

  /**
   * Extract metadata directives from genome source comments.
   *
   * Supported directives:
   * - `; @mode: bidirectional` - Centered mode, value 32 = no movement (default)
   * - `; @mode: forward-only` - Legacy mode, value 0 = no movement
   *
   * @param source - Raw genome string
   * @returns Extracted metadata with defaults for unspecified values
   */
  private extractDirectives(source: string): GenomeMetadata {
    const metadata: GenomeMetadata = { ...DEFAULT_GENOME_METADATA };
    const lines = source.split("\n");

    for (const line of lines) {
      const commentIdx = line.indexOf(";");
      if (commentIdx < 0) continue;

      const comment = line.slice(commentIdx + 1).trim();

      // Parse @mode directive - accepts user-friendly and internal names
      const modeMatch = comment.match(
        /^@mode:\s*(bidirectional|forward-only|centered|forward)$/i,
      );
      if (modeMatch) {
        const userMode = modeMatch[1].toLowerCase();
        metadata.mode = MODE_ALIASES[userMode] ?? "centered";
      }
    }

    return metadata;
  }

  private validateLineFrame(
    line: string,
    lineIdx: number,
    errors: ParseError[],
  ) {
    const codeLine = getCodeContent(line);
    let baseCount = 0;

    for (let charIdx = 0; charIdx < codeLine.length; charIdx++) {
      const char = codeLine[charIdx];
      if (char === undefined) continue;
      if (this.validBases.has(char)) {
        baseCount++;
      } else if (char.trim() === "") {
        // Whitespace detected
        if (baseCount % 3 !== 0) {
          errors.push({
            message: `Mid-triplet break detected at line ${lineIdx + 1}. ${
              baseCount % 3
            } base(s) before whitespace.`,
            position: charIdx,
            severity: "warning",
            fix: "Remove whitespace or complete the codon",
          });
        }
      }
    }
  }

  private checkStartCodon(tokens: CodonToken[], errors: ParseError[]) {
    const firstToken = tokens[0];
    if (firstToken && firstToken.text !== "ATG") {
      errors.push({
        message: "Program should begin with START codon (ATG)",
        position: 0,
        severity: "error",
        fix: "Add ATG at the beginning",
      });
    }
  }

  private checkStopCodons(tokens: CodonToken[], errors: ParseError[]) {
    const stopCodons = new Set(["TAA", "TAG", "TGA"]);
    let firstStopIdx = -1;

    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];
      if (!token) continue;

      if (stopCodons.has(token.text)) {
        if (firstStopIdx === -1) {
          firstStopIdx = i;
        }
      } else if (firstStopIdx >= 0 && token.text === "ATG") {
        // START after STOP
        errors.push({
          message: `START codon after STOP at position ${i}`,
          position: token.position,
          severity: "warning",
          fix: "Remove unreachable code after STOP",
        });
      }
    }

    // Check if program ends with STOP
    const lastToken = tokens[tokens.length - 1];
    if (lastToken && !stopCodons.has(lastToken.text)) {
      errors.push({
        message: "Program should end with STOP codon (TAA, TAG, or TGA)",
        position: lastToken.position,
        severity: "warning",
        fix: "Add TAA at the end",
      });
    }
  }
}

File: engine/core/src/lexer.rs
//! Tokenizer and structural validation for genome source.
//!
//! Parsing rules (unchanged from the original):
//! - Valid bases are `A C G T U`; `U` is normalized to `T`.
//! - Whitespace is ignored and may be used for formatting.
//! - `;` starts a comment that runs to end of line.
//! - The stripped source length must be divisible by 3.

use serde::Serialize;

use crate::base::{is_base, normalize_base};
use crate::codon::{is_stop, opcode_for};

/// A single parsed codon with its source position.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Token {
    /// The three-letter codon, normalized to DNA notation.
    pub text: String,
    /// Index of the codon's first base within the stripped source.
    pub position: usize,
    /// 1-based source line the codon starts on.
    pub line: usize,
}

/// Severity of a [`Diagnostic`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Error,
    Warning,
}

/// A structural or frame-alignment finding. Warnings never block execution.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Diagnostic {
    pub message: String,
    pub position: usize,
    pub severity: Severity,
    pub fix: String,
}

impl Diagnostic {
    fn warning(message: impl Into<String>, position: usize, fix: impl Into<String>) -> Self {
        Self {
            message: message.into(),
            position,
            severity: Severity::Warning,
            fix: fix.into(),
        }
    }
    fn error(message: impl Into<String>, position: usize, fix: impl Into<String>) -> Self {
        Self {
            message: message.into(),
            position,
            severity: Severity::Error,
            fix: fix.into(),
        }
    }
}

/// Strips a trailing `;` comment from a line.
fn code_content(line: &str) -> &str {
    match line.find(';') {
        Some(i) => &line[..i],
        None => line,
    }
}

/// Tokenizes genome source into codons.
///
/// Returns `Err` with a human-readable message for hard errors (invalid base
/// characters, or a stripped length that is not a multiple of three), matching
/// the exceptions thrown by the original lexer.
pub fn tokenize(source: &str) -> Result<Vec<Token>, String> {
    let mut cleaned = String::new();
    // Line each cleaned base came from, parallel to `cleaned`.
    let mut line_of_base: Vec<usize> = Vec::new();

    for (line_idx, line) in source.split('\n').enumerate() {
        for (col, ch) in code_content(line).chars().enumerate() {
            let upper = ch.to_ascii_uppercase();
            if is_base(upper) {
                cleaned.push(normalize_base(upper));
                line_of_base.push(line_idx + 1);
            } else if !ch.is_whitespace() {
                return Err(format!(
                    "Invalid character '{ch}' at line {}, column {col}",
                    line_idx + 1
                ));
            }
        }
    }

    if !cleaned.len().is_multiple_of(3) {
        let missing = 3 - (cleaned.len() % 3);
        return Err(format!(
            "Source length {} is not divisible by 3. Missing {missing} bases for complete codon.",
            cleaned.len()
        ));
    }

    let bytes = cleaned.as_bytes();
    let mut tokens = Vec::with_capacity(cleaned.len() / 3);
    let mut i = 0;
    while i < bytes.len() {
        // `cleaned` is ASCII, so byte and char indices coincide.
        let text = cleaned[i..i + 3].to_string();
        tokens.push(Token {
            text,
            position: i,
            line: line_of_base.get(i).copied().unwrap_or(1),
        });
        i += 3;
    }
    Ok(tokens)
}

/// Validates reading-frame alignment, reporting whitespace breaks that fall
/// mid-triplet. Returns warnings only; never errors.
pub fn validate_frame(source: &str) -> Vec<Diagnostic> {
    let mut diags = Vec::new();
    for (line_idx, line) in source.split('\n').enumerate() {
        let mut base_count = 0usize;
        for (col, ch) in code_content(line).chars().enumerate() {
            if is_base(ch.to_ascii_uppercase()) {
                base_count += 1;
            } else if ch.is_whitespace() && !base_count.is_multiple_of(3) {
                diags.push(Diagnostic::warning(
                    format!(
                        "Mid-triplet break detected at line {}. {} base(s) before whitespace.",
                        line_idx + 1,
                        base_count % 3
                    ),
                    col,
                    "Remove whitespace or complete the codon",
                ));
            }
        }
    }
    diags
}

/// Validates structural integrity: START at the beginning, a STOP at the end,
/// and no START after a STOP. Returns errors and warnings.
pub fn validate_structure(tokens: &[Token]) -> Vec<Diagnostic> {
    let mut diags = Vec::new();
    if tokens.is_empty() {
        return diags;
    }

    if tokens[0].text != "ATG" {
        diags.push(Diagnostic::error(
            "Program should begin with START codon (ATG)",
            0,
            "Add ATG at the beginning",
        ));
    }

    let mut first_stop: Option<usize> = None;
    for (i, token) in tokens.iter().enumerate() {
        if is_stop(&token.text) {
            first_stop.get_or_insert(i);
        } else if first_stop.is_some() && token.text == "ATG" {
            diags.push(Diagnostic::warning(
                format!("START codon after STOP at position {i}"),
                token.position,
                "Remove unreachable code after STOP",
            ));
        }
    }

    if let Some(last) = tokens.last()
        && !is_stop(&last.text)
    {
        diags.push(Diagnostic::warning(
            "Program should end with STOP codon (TAA, TAG, or TGA)",
            last.position,
            "Add TAA at the end",
        ));
    }

    diags
}

/// Reports unknown codons (codons with no opcode). The standard map covers all
/// 64 codons, so this only fires on malformed tokens.
pub fn unknown_codons(tokens: &[Token]) -> Vec<Diagnostic> {
    tokens
        .iter()
        .filter(|t| opcode_for(&t.text).is_none())
        .map(|t| {
            Diagnostic::error(
                format!("Unknown codon '{}'", t.text),
                t.position,
                "Replace with a valid codon",
            )
        })
        .collect()
}
